### PR TITLE
Feat/governance and timelock tests

### PR DESCRIPTION
**Pull request by baroooo**
### Description

To verify the deployed Mento Govenor & Timelock contracts work as expected we need to create fork tests, as defined in the test plan, to the existing suite of tests.

- [x]  Create a proposal
    - [x]  Create a proposal to transfer some tokens from the treasury
    - [x]  Vote on the proposal using multiple accounts, with the majority voting **YES**
    - [x]  Queue the proposal
    - [x]  Execute the proposal
    - [x]  Tokens should be transferred to the target account
- [x]  Cancel the proposal
    - [x]  Create another proposal to transfer tokens from the treasury
    - [x]  Vote on the proposal using multiple accounts, with the majority voting **YES**
    - [x]  Queue the proposal
    - [x]  Cancel the proposal using the watchdog account
    - [x]  The proposal should be cancelled and can not be executed
- [x]  Defeat a proposal
    - [x]  Create another proposal
    - [x]  Vote on the proposal using multiple accounts, with the majority voting **NO**
    - [x]  The proposal should be defeated and can not be executed
- [x]  Change Mento Governor config
    - [x]  Create a proposal to update the governor config
        - [x]  `votingDelay`
        - [x]  `votingPeriod`
        - [x]  `threshold`
        - [x]  `quorum`
        - [x]  `minDelay`
        - [x]  `minCliffPeriod`
        - [x]  `minSlopePeriod`
    - [x]  Vote on and execute the proposal
    - [x]  Config is updated, and changes are effective
- [x]  Change roles on the Governance Timelock
    - [x]  Create a proposal to:
        - [x]  Remove the current proposer, executor, and canceller
        - [x]  Add new addresses with these roles
    - [x]  Ensure the new addresses have the correct roles
- [x]  Update upgradeable contracts
    - [x]  Create and deploy new implementations of
        - [x]  Locking
        - [x]  Governance Timelock
        - [x]  Mento Governor
    - [x]  Create a proposal to update the contracts’ implementation
    - [x]  Vote on and execute the proposal
    - [x]  All upgradeable contracts are updated, and changes are verified

### Other changes

- Removed node creation since test runner does that already
- Mock implementations to simulate upgrades
- Time travel moves timestamp and block number in sync
- Import OZ's proxy admin and compile contracts first to create OpenZeppelin types

### Related issues

- #8

### CODE REVIEW

File: test/Governance.test.ts [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+    // Create a proposal to update governor config
+    const proposalId = submitProposal(
+      alice,
+      targets,
+      values,
+      calldatas,
+      description,
+    );
+
+    // Vote on the proposal using multiple accounts, with the majority voting NO
```
Comment (philbow61) [inline test/Governance.test.ts:306]: YES 😁

File: test/Governance.test.ts [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+    // Create a proposal to change governor roles
+    const proposalId = submitProposal(
+      alice,
+      targets,
+      values,
+      calldatas,
+      description,
+    );
+
+    // Vote on the proposal using multiple accounts, with the majority voting NO
```
Comment (philbow61) [inline test/Governance.test.ts:383]: YES 😁

File: test/Governance.test.ts [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+    // Create a proposal to upgrade contracts
+    const proposalId = submitProposal(
+      alice,
+      targets,
+      values,
+      calldatas,
+      description,
+    );
+
+    // Vote on the proposal using multiple accounts, with the majority voting NO
```
Comment (philbow61) [inline test/Governance.test.ts:453]: YES 😁

File: hardhat.config.ts [hunk 4]
```diff
@@ -29,6 +29,15 @@ const config: HardhatUserConfig = {
...
     },
   },
+  solidity: {
+    version: '0.8.18',
+    settings: {
+      optimizer: {
+        enabled: true,
+        runs: 200,
+      },
+    },
```
Comment (philbow61) [inline hardhat.config.ts:39]: id did add viaIR: true, in my. pr depending on who merges first we need to check

**Review by philbow61 — APPROVED** [review]
LGTM 👍

File: test/Governance.test.ts (hunk 1 above)
Comment (baroooo) [reply test/Governance.test.ts:306]: 🤦

File: contracts/Mocks/MockLocking.sol [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+contract MockLocking {
+  function getWeek(
+  ) external pure {
+    revert("MockLocking: lock not implemented");
```
Comment (nvtaveras) [inline contracts/Mocks/MockLocking.sol:7]: `getWeek` not implemented?

**Review by nvtaveras — APPROVED** [review]
Looks great! 🥇

File: test/Governance.test.ts [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+  before(async function () {
+    const chainId = await setupEnvironment();
+    console.log('\r\n========================');
+    console.log('Running Governance tests on network with chain id:', chainId);
+    console.log('========================\r\n');
+  });
+
+  beforeEach(async function () {
+    // reset the fork state between tests to not pollute the state
+    // @ts-expect-error - forking doesn't exist in hre for some reason
```
Comment (chapati23) [inline test/Governance.test.ts:60]:
```suggestion
    // @ts-expect-error - forking doesn't exist in here for some reason
```

File: test/Governance.test.ts [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+      `AccessControl: account ${alice.address.toLowerCase()} is missing role ${cancellerRole}`,
+    );
+
+    const watchdogAddress = await governanceFactory.watchdogMultiSig();
+    const watchdog = await getImpersonatedSigner(watchdogAddress);
+
+    expect(await timelock.isOperationPending(timelockId)).to.be.true;
+    // watchdog can cancel the proposal
+    await timelock.connect(watchdog).cancel(timelockId);
+    expect(await timelock.isOperationPending(timelockId)).to.be.false;
```
Comment (chapati23) [inline test/Governance.test.ts:193]: should we also check that the proposal itself has status cancelled now with `expect(await governor.state(proposalId)).to.eq(2)`?

File: test/Governance.test.ts [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+
+    expect(await timelock.hasRole(proposerRole, newProposer)).to.be.true;
+    expect(await timelock.hasRole(cancellerRole, newCanceller)).to.be.true;
+    expect(
+      await timelock.hasRole(proposerRole, governanceAddresses.MentoGovernor),
+    ).to.be.false;
+    expect(await timelock.hasRole(cancellerRole, watchdogAddress)).to.be.false;
+  });
+
+  it('should upgrade upgradable contracts', async function () {
```
Comment (chapati23) [inline test/Governance.test.ts:413]: just a note that we'll need to include the `Emissions` contract in this test as well after https://github.com/mento-protocol/mento-core/pull/365 is merged

leave it up to you if you want to do it in this PR or create a new issue

**Review by chapati23 — APPROVED** [review]
great work! lgtm

**Pull request merged**